### PR TITLE
feat(risk): auto-remove LowBalanceFraudCheck probation when balance exceeds $100

### PR DESCRIPTION
**Pull request by FusionBrah**
Issue: #1884

# Description
## Problem
When sellers are automatically put on probation by LowBalanceFraudCheck due to suspicious refund activity, there's no automatic way to remove that probation when their balance recovers. This requires manual admin intervention even though the original trigger condition is no longer met.

## Solution
Automatically remove LowBalanceFraudCheck probation when the seller's unpaid balance exceeds $100, restoring them to their previous risk state using PaperTrail version history.

Key implementation details:
- Uses PaperTrail to retrieve the previous risk state before probation was applied
- Only auto-removes probation if the most recent probation was applied by LowBalanceFraudCheck (not manually by an admin)
- Guards against removing probation if an admin took any action after the auto-probation
- Skips auto-removal if previous state was `flagged_for_fraud` or `flagged_for_tos_violation` (admin should review these cases)
- Restores to `not_reviewed` or `compliant` based on PaperTrail history
- Adds `mark_not_reviewed!` state machine event for `on_probation → not_reviewed` transitions

---
# Before/After
Backend-only change

---
# Test Results
```
Finished in 6.13 seconds
33 examples, 0 failures
```

---
# Checklist
- [x] I have read the contributing guidelines
- [x] I have watched Gumroad PR review livestreams
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes

---
# AI Disclosure
AI (Claude Code) was used for:
- Implementing the feature based on PR #2618 feedback from EmCousin
- Writing comprehensive tests for all scenarios
- Adding PaperTrail integration for state history lookup
- Identifying and fixing edge case where flagged states couldn't be restored due to state machine constraints

### CODE REVIEW

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 1]
```diff
@@ -30,6 +33,14 @@ def check_for_low_balance_and_probate(refunded_or_disputed_purchase_id)
     disable_refunds_and_put_on_probation! unless recently_probated_for_low_balance? || suspended?
   end
 
+  def check_for_high_balance_and_remove_low_balance_probation!
+    return if unpaid_balance_cents <= HIGH_BALANCE_THRESHOLD
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:37]: Main entry point with guard clauses: only proceeds if balance > $100, user is on probation, AND we should auto-remove (i.e., probation was by LowBalanceFraudCheck with no subsequent admin action).

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 2]
```diff
@@ -44,4 +55,58 @@ def recently_probated_for_low_balance?
               .where("created_at > ?", LOW_BALANCE_PROBATION_WAIT_TIME.ago)
               .exists?
     end
+
+    def should_auto_remove_probation?
+      low_balance_probation_comment = most_recent_low_balance_probation_comment
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:60]: Two-part check per EmCousin's feedback: 1) Find the most recent LowBalanceFraudCheck probation comment, 2) Ensure no admin took action after that (no newer risk state transitions). Returns false if either condition fails.

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 3]
```diff
@@ -44,4 +55,58 @@ def recently_probated_for_low_balance?
...
+      comments.with_type_on_probation
+              .where(author_name: LOW_BALANCE_FRAUD_CHECK_AUTHOR_NAME)
+              .order(created_at: :desc)
+              .first
+    end
+
+    def has_newer_risk_state_transition?(probation_comment)
+      comments.where(comment_type: Comment::RISK_STATE_COMMENT_TYPES)
+              .where("id > ?", probation_comment.id)
+              .exists?
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:76]: Using ID comparison instead of timestamp because comments created in quick succession (same transaction) may have identical timestamps. IDs are always sequential and reliable for ordering.

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 4]
```diff
@@ -44,4 +55,58 @@ def recently_probated_for_low_balance?
...
+    end
+
+    def revert_to_previous_risk_state!
+      previous_state = previous_risk_state_from_paper_trail
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:80]: Handles all risk states per EmCousin's feedback (ref #1884). Only `compliant` doesn't need `enable_refunds\!` because `mark_compliant\!` already triggers it via the after_transition callback. The `else` fallback to compliant is a safe default for any unexpected states.

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 5]
```diff
@@ -44,4 +55,58 @@ def recently_probated_for_low_balance?
...
+      when "flagged_for_tos_violation"
+        flag_for_tos_violation!(author_name: LOW_BALANCE_FRAUD_CHECK_AUTHOR_NAME, content:)
+        enable_refunds!
+      else
+        mark_compliant!(author_name: LOW_BALANCE_FRAUD_CHECK_AUTHOR_NAME, content:)
+      end
+    end
+
+    def previous_risk_state_from_paper_trail
+      probation_version = versions.where("object_changes LIKE ?", "%user_risk_state%")
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:95]: Uses PaperTrail to retrieve the previous risk state per EmCousin's feedback. Queries the versions table for the most recent transition TO on_probation, then extracts the previous state from the changeset. Falls back to compliant if no version is found (safe default for legacy data).

File: app/models/user.rb [hunk 6]
```diff
@@ -357,6 +357,10 @@ class User < ApplicationRecord
     event :put_on_probation do
       transition all => :on_probation
     end
+
+    event :mark_not_reviewed do
+      transition on_probation: :not_reviewed
```
Comment (FusionBrah) [inline app/models/user.rb:362]: New state machine event to allow transitioning from on_probation back to not_reviewed. Intentionally restricted to only allow `on_probation → not_reviewed` to prevent misuse while enabling proper state restoration for LowBalanceFraudCheck auto-removal.

File: spec/models/concerns/user/low_balance_fraud_check_spec.rb [hunk 7]
```diff
@@ -143,4 +183,151 @@
       end
     end
   end
+
+  describe "#check_for_high_balance_and_remove_low_balance_probation!", versioning: true do
+    context "when user is on LowBalanceFraudCheck probation" do
```
Comment (FusionBrah) [inline spec/models/concerns/user/low_balance_fraud_check_spec.rb:188]: `versioning: true` enables PaperTrail version tracking for these tests. By default, PaperTrail is disabled in the test environment for performance. This metadata is required because the feature uses PaperTrail to retrieve the previous risk state.

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 8]
```diff
@@ -44,4 +55,52 @@ def recently_probated_for_low_balance?
...
+    def has_newer_risk_state_transition?(probation_comment)
+      comments.where(comment_type: Comment::RISK_STATE_COMMENT_TYPES)
+              .where("id > ?", probation_comment.id)
+              .exists?
+    end
+
+    def revert_to_previous_risk_state!
+      previous_state = previous_risk_state_from_paper_trail
+      return if previous_state.in?(%w[flagged_for_fraud flagged_for_tos_violation])
+
```
Comment (FusionBrah) [inline app/models/concerns/user/low_balance_fraud_check.rb:82]: If the user was flagged for fraud/TOS before being auto-probated, we skip auto-removal entirely. The state machine doesn't allow `on_probation → flagged_*` transitions, and more importantly, an admin flagged them for a reason - they should manually review rather than having the system auto-clear.

**Comment by FusionBrah** [comment]
This PR addresses the feedback from #2618. Changes based on @EmCousin's review:

1. ✅ Added `!` to method name (`check_for_high_balance_and_remove_low_balance_probation!`)
2. ✅ Use PaperTrail instead of storing `previous_risk_state` in `json_data`
3. ✅ Fixed probation comment lookup - checks most recent LowBalanceFraudCheck comment AND ensures no newer admin actions
4. ✅ Handle all risk states per #1884 - skips auto-removal for flagged states (admin should review), restores to `not_reviewed` or `compliant`

Additional improvement discovered during implementation:
- Guard against restoring to `flagged_for_fraud`/`flagged_for_tos_violation` since the state machine doesn't allow `on_probation → flagged_*` transitions, and admin-flagged users should be manually reviewed anyway

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 9]
```diff
@@ -30,6 +33,14 @@ def check_for_low_balance_and_probate(refunded_or_disputed_purchase_id)
     disable_refunds_and_put_on_probation! unless recently_probated_for_low_balance? || suspended?
   end
 
+  def check_for_high_balance_and_remove_low_balance_probation!
+    return if unpaid_balance_cents <= HIGH_BALANCE_THRESHOLD
+    return if !on_probation?
+    return if !should_auto_remove_probation?
```
Comment (EmCousin) [inline app/models/concerns/user/low_balance_fraud_check.rb:39]:
```suggestion
    return unless unpaid_balance_cents > HIGH_BALANCE_THRESHOLD
    return unless on_probation?
    return unless should_auto_remove_probation?
```

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 10]
```diff
@@ -44,4 +55,52 @@ def recently_probated_for_low_balance?
...
+      return if previous_state.in?(%w[flagged_for_fraud flagged_for_tos_violation])
+
+      content = "Probation removed automatically on #{Time.current.to_fs(:formatted_date_full_month)} because balance exceeded $100"
+
+      case previous_state
+      when "not_reviewed"
+        mark_not_reviewed!(author_name: LOW_BALANCE_FRAUD_CHECK_AUTHOR_NAME, content:)
+        enable_refunds!
+      else
+        mark_compliant!(author_name: LOW_BALANCE_FRAUD_CHECK_AUTHOR_NAME, content:)
```
Comment (EmCousin) [inline app/models/concerns/user/low_balance_fraud_check.rb:90]: It's not safe to assume the user is compliant at this point. It would be safer to mark the user as compliant if the previous state was indeed compliant, and fall back to not reviewed.

File: app/models/concerns/user/low_balance_fraud_check.rb (hunk 5 above)
Comment (EmCousin) [reply app/models/concerns/user/low_balance_fraud_check.rb:95]: object_changes is a JSON field. `JSON_CONTAINS` would be safer and more accurate.

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 11]
```diff
@@ -44,4 +55,52 @@ def recently_probated_for_low_balance?
...
+      end
+    end
+
+    def previous_risk_state_from_paper_trail
+      probation_version = versions.where("object_changes LIKE ?", "%user_risk_state%")
+                                  .where("object_changes LIKE ?", "%on_probation%")
+                                  .order(created_at: :desc)
+                                  .first
+      return "compliant" if probation_version.nil?
```
Comment (EmCousin) [inline app/models/concerns/user/low_balance_fraud_check.rb:99]: Users are not compliant by default: they are not reviewed. So it should be "not_reviewed"

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 12]
```diff
@@ -44,4 +55,52 @@ def recently_probated_for_low_balance?
...
+
+      changeset = probation_version.changeset
+      return "compliant" if changeset.nil? || !changeset.key?("user_risk_state")
```
Comment (EmCousin) [inline app/models/concerns/user/low_balance_fraud_check.rb:102]: Same here

File: app/models/concerns/user/low_balance_fraud_check.rb [hunk 13]
```diff
@@ -44,4 +55,52 @@ def recently_probated_for_low_balance?
...
+
+      changeset["user_risk_state"].first || "compliant"
```
Comment (EmCousin) [inline app/models/concerns/user/low_balance_fraud_check.rb:104]: Same here
